### PR TITLE
Fix vtable chunk sharing with covariant returns

### LINKED ISSUE #42047: Covariant returns support has a bug in edge cases
We incorrectly end up sharing a chunk of vtable between a parent class and a child class that uses a covariant override of a method that occupies one of the previous vtable chunks. We end up modifying a slot in the parent class vtable.
In a simple testing application, it leads to a stack overflow caused by an infinite recursion in the type system code.

### PR DESCRIPTION
**Pull request by janvorli**
There is an issue causing invalid sharing of a vtable chunk when a
method is overridden using method impl in a class and its parent class
and the slot in the parent class is in a previous vtable chunk. The
algorithm in `MethodTableBuilder::ChangesImplementationOfVirtualSlot` that
detects whether a slot in a vtable chunk was changed doesn't know that
the override will be propagated to all slots occupied by the method
being overridden. In the test attached to this fix, it leads to a stack
overflow due to an infinite recursion in the type system code. It could
also result in an overwrite of the method in the parent class, so when
another class derives from it, it would get a method from its sibling.

This change fixes it by checking if the parent method of a current slot
was overridden in some previous slot in the vtable.

This is a fix for #42047

### CODE REVIEW

File: src/coreclr/src/vm/methodtablebuilder.cpp [hunk 1]
```diff
@@ -11602,6 +11602,23 @@ BOOL MethodTableBuilder::ChangesImplementationOfVirtualSlot(SLOT_INDEX idx)
...
             fChangesImplementation = TRUE;
+
+        if (!fChangesImplementation)
+        {
+            // Search the previous slots in the parent vtable for the same implementation. If it exists and it was
+            // overriden, the ClassLoader::PropagateCovariantReturnMethodImplSlots will propagate the change to the current
+            // slot (idx), so the implementation of it will change.
+
+            MethodDesc* pParentMD = ParentImpl.GetMethodDesc();
+            for (SLOT_INDEX i = 0; i < idx; i++)
```
Comment (janvorli) [inline src/coreclr/src/vm/methodtablebuilder.cpp:11613]: @davidwrighton one thing I was considering here is to run the lookup loop only when the PreserveBaseOverridesAttribute is found on the pParentMD. But I was not sure what is the performance profile of the `MethodDesc::GetCustomAttribute(WellKnownAttribute::PreserveBaseOverridesAttribute, ...)`.

Comment (janvorli) [reply src/coreclr/src/vm/methodtablebuilder.cpp:11613]: Ah, I haven't realized before that I could call the RequiresCovariantReturnTypeChecking on the pParentMD and that check is cheap.

Comment (janvorli) [reply src/coreclr/src/vm/methodtablebuilder.cpp]: I've added commit with such a change.

File: src/coreclr/src/vm/methodtablebuilder.cpp [hunk 2]
```diff
@@ -11602,6 +11602,23 @@ BOOL MethodTableBuilder::ChangesImplementationOfVirtualSlot(SLOT_INDEX idx)
         // methods.
         if (!fChangesImplementation && (ParentImpl.GetSlotIndex() != idx))
             fChangesImplementation = TRUE;
+
+        if (!fChangesImplementation)
+        {
+            // Search the previous slots in the parent vtable for the same implementation. If it exists and it was
+            // overriden, the ClassLoader::PropagateCovariantReturnMethodImplSlots will propagate the change to the current
+            // slot (idx), so the implementation of it will change.
+
```
Comment (davidwrighton) [inline src/coreclr/src/vm/methodtablebuilder.cpp]: While I believe this is correct, this algorithm is O(n^2) as ChangesImplementationOfVirtualSlot is called for each vtable slot inherited from the parent type. Particular concerns arise from WPF and WinForms scenarios as many of the control classes have >>100 virtual methods.

I believe we either need to 
1. Don't run this algorithm when propagation will not occur. For instance, we have the `HasVTableMethodImpl` check in `PropagateCovariantReturnMethodImplSlots` and checks for PreserveBaseOverridesAttribute before running the secondary loop that makes it O(N^2).
OR 
2. Make the algorithm here not O(N^2)

**Review by davidwrighton — CHANGES_REQUESTED** [review]

File: src/coreclr/src/vm/methodtablebuilder.cpp [hunk 3]
```diff
@@ -11605,17 +11605,20 @@ BOOL MethodTableBuilder::ChangesImplementationOfVirtualSlot(SLOT_INDEX idx)
 
         if (!fChangesImplementation)
         {
-            // Search the previous slots in the parent vtable for the same implementation. If it exists and it was
-            // overriden, the ClassLoader::PropagateCovariantReturnMethodImplSlots will propagate the change to the current
-            // slot (idx), so the implementation of it will change.
-
             MethodDesc* pParentMD = ParentImpl.GetMethodDesc();
-            for (SLOT_INDEX i = 0; i < idx; i++)
+            if (pParentMD->RequiresCovariantReturnTypeChecking())
```
Comment (davidwrighton) [inline src/coreclr/src/vm/methodtablebuilder.cpp:11609]: This isn't correct. The propagation feature isn't actually tied to covariant return type checking, and I probably should have caught that in the previous change, as the PropagateCovariantReturnMethodImplSlots name isn't really right.

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**